### DISCUSSION TITLE
accessing material property from interface kernel

### DISCUSSION THREAD

**Discussion by schakrabortygithub** [idaholab/moose#26372] [Q&A General]
Hi,
I am trying to write an interface kernel for which I need the material property values inside the element not at the element faces. Similarly for the neighbor side.

Based on my understanding, when we try to access any material property using '_MyProperty(getMaterialProperty<TYPE>("property_name"))' it provides the corresponding material property based on the consuming object. If the consuming object is a 'Kernel' then it is evaluated at the present working integration points i.e, inside the element. If it is a boundary kernel then at the face of the element.

How can I access the material property(volume averaged) at the inside of element while operating on a boundary or interface kernel?

Best
Subhendu

**Comment by GiudGiud** [comment]
hello

So when you say inside the element, where inside the element? At the center?
When we do element integration loops (kernels) we compute material properties at the quadrature points in the element. Similarly here, the interface kernel loop is looking at quadrature points on the face of the element on the boundary of interest. 

Guillaume

**Reply by lindsayad** [reply]
It sounds like he wants volume averaged. We don't support this in a "standard" way, but you could loosely achieve what you want by writing your own object, using the [Functor system](https://mooseframework.inl.gov/syntax/Functors/index.html), and evaluating the functors with `ElemArg`. The `ElemArg` result would likely not represent a volume-average though; more like a centroid evaluation

**Reply by GiudGiud** [reply]
with a little templating logic, one could write the `ElemArg` to be a volume averaging of the material property. I've done this sort of the 'special' behavior for certain functor arguments in the FunctorSmoother.

I agree using the Functor system to write a functor material property instead of a regular material property will let you do this sort of non/less-local evaluations much easier

**Reply by schakrabortygithub** [reply]
Hi @GiudGiud and @lindsayad ,

Just to clarify the problem, I am working on a kernel and interface_kernel where material properties (second_piola_kirchhoff_stress) are obtained from the class 'ComputeMultipleCrystalPlasticityStress'. That class seems not to support the computation of material property for element-faces(_bnd=1) or face-neighbor(_neighbor=1).

https://github.com/idaholab/moose/blob/c3a71595d3d6151111d0a65bdc32d8cfe11b57c1/modules/tensor_mechanics/src/materials/crystal_plasticity/ComputeMultipleCrystalPlasticityStress.C#L201C1-L204C1

I was wondering if while operating on the interface kernel, I can access those properties at the element gauss points (not on element-face GPs) so that I can compute the volume average value of that property on that element and use it as approximation for the element faces.

**Reply by lindsayad** [reply]
Hmm that's a pretty complex material class. I'm not sure we'd want to write a `FunctorMaterial` equivalent for it without first considering other options. Perhaps a first question would be why only volumetric material calculations are supported for that class. @dewenyushu @sapitts ?

**Reply by dewenyushu** [reply]
We (at least myself) have not met cases where other types of material property (in addition to volumetric) is required in the crystal plasticity material classes. I do not see obvious reasons why we should not support it, though.

**Reply by sapitts** [reply]
I agree with @dewenyushu about no obvious reason not to support the interface options in the crystal plasticity class. I suspect that the error is a hold over from the older user object implementation. There might be some hidden reason (e.g. grain boundary mechanics) lurking in the background, but I don't think that's a showstopper--just a corner case that would need to be properly supported.

**Reply by lindsayad** [reply]
So @schakrabortygithub I would try removing that early return in the material class and see if you can use the face-qp material property evaluations in your interface kernel

**Reply by schakrabortygithub** [reply]
@dewenyushu  Here is one example where we need those material properties from the boundary and interface kernel:
We are developing a dislocation transport based crystal plasticity material model. In this model, dislocations evolve via transport (advection kind of model) within the grain and also there is transfer across the Grain Boundary(GB). Among others, there's a material class inherited from 'CrystalPlasticityStressUpdateBase' that compute the dislocation velocity vector which needs resolved shear stress as input. The flux across the GB depends on the local resolved shear stress via dislocation velocity. This transfer across GB is implemented as an interface kernel. That's where we need those material properties for the element-face and face-neighbor. We can always approximate those values with volumetric value in the current element and the neighbor element of the current element-face as long as those material properties are accessible within the interface kind of kernel.

**Reply by schakrabortygithub** [reply]
@lindsayad that will be good.
I also tried to remove that early return and let those properties get computed for _bnd and _neighbor as well. At the onset of plasticity, volumetric stress converges but stress update loop for face throws an error "ComputeMultipleCrystalPlasticityStress: Constitutive failure".
